src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs

```diff
@@ -65,14 +65,28 @@ internal static SafeFreeContextBuffer CreateEmptyHandle()
             return new SafeFreeContextBuffer_SECURITY();
         }
 
-        //
-        // After PInvoke call the method will fix the refHandle.handle with the returned value.
-        // The caller is responsible for creating a correct SafeHandle template or null can be passed if no handle is returned.
-        //
-        // This method switches between three non-interruptible helper methods.  (This method can't be both non-interruptible and
-        // reference imports from all three DLLs - doing so would cause all three DLLs to try to be bound to.)
-        //
-        public static unsafe int QueryContextAttributes(SafeDeleteContext phContext, Interop.SspiCli.ContextAttribute contextAttribute, byte* buffer, SafeHandle? refHandle)
+        public static unsafe int QueryContextAttributes(SafeDeleteContext phContext, Interop.SspiCli.ContextAttribute contextAttribute, IntPtr* handle)
+        {
+            try
+            {
+                bool ignore = false;
+                phContext.DangerousAddRef(ref ignore);
+                return Interop.SspiCli.QueryContextAttributesW(ref phContext._handle, contextAttribute, handle);
+            }
+            finally
+            {
+                phContext.DangerousRelease();
+            }
```

Comment:
I know this pattern is used on many places, but shouldn't it be something like

```suggestion
            bool mustRelease = false;
            try
            {
                phContext.DangerousAddRef(ref mustRelease);
                return Interop.SspiCli.QueryContextAttributesW(ref phContext._handle, contextAttribute, handle);
            }
            finally
            {
                if (mustRelease)
                {
                    phContext.DangerousRelease();
                }
            }
```

Reply:
> I know this pattern is used on many places,

It is?? Where?

Reply:
I see about 6 occurences in this file, then also SSPIAuthType.cs

Fulltext search for DangerousAddRef(ref ignore) spits out much.

Reply:
It's valid if you're 100% sure by construction that the SafeHandle is not disposed nor referenced by any other code. I hope that's what those other occurrences are; if not, they must be fixed. From a quick glance, I don't see anything local to this function that guarantees that invariant.

Reply:
Sure. I will look into that, since most of those seem to be in the SslStream-related areas.

Reply:
I raised https://github.com/dotnet/runtime/pull/97394. @wfurt, there will be a conflict when either one is merged. feel free to merge this one first and I will resolve the conflicts in my PR.

Reply:
updated.